runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,346 @@
+          in_braces = false
+        elseif c == "[" then
+          in_brackets = true
+        elseif c == "]" then
+          in_brackets = false
+        end
+        segment = segment..c
+      end
+    end
+    if segment ~= "" then table.insert(segments, segment) end
+    return segments
+  end
+
+  local function escape(c)
+    if c == "?" or c == "." or c == "(" or c == ")" or c == "%" or
+       c == "[" or c == "]" or c == "*" or c == "+" or c == "-" then
+      return "%"..c
+    end
+    return c
+  end
+
+  local segments = split(pattern, "/")
+  for i, segment in ipairs(segments) do
+    local last_seg = i == #segments
+    if segment == "**" then
+      local chunks = {
+        path_sep.."-",
+        ".-"..path_sep,
+      }
+      if last_seg then
+        chunks = { ".-" }
+      end
+      append(chunks)
+    else
+      local in_braces = false
+      local brace_val = ""
+      local in_brackets = false
+      local bracket_val = ""
+      for j = 1, #segment do
+        local char = string.sub(segment, j, j)
+        if char ~= "}" and in_braces then
+          brace_val = brace_val..char
+        else
+          if in_brackets and (char ~= "]" or bracket_val == "") then
+            local res
+            if char == "-" then
+              res = char
+            elseif bracket_val ~= "" and char == "!" then
+              res = "^"
+            elseif char == "**" then
+              res = ""
+            else
+              res = escape(char)
+            end
+            bracket_val = bracket_val..res
+          else
+            if char == "{" then
+              in_braces = true
+            elseif char == "[" then
+              in_brackets = true
+            elseif char == "}" then
+              local choices = split(brace_val, ",")
+              local parsed_choices = {}
+              for _, choice in ipairs(choices) do
+                table.insert(parsed_choices, parse(choice))
+              end
+              append(vim.tbl_flatten(parsed_choices))
+              in_braces = false
+              brace_val = ""
+            elseif char == "]" then
+              append({"["..bracket_val.."]"})
+              in_brackets = false
+              bracket_val = ""
+            elseif char == "?" then
+              append({non_path_sep})
+            elseif char == "*" then
+              append({non_path_sep.."-"})
+            else
+              append({escape(char)})
+            end
+          end
+        end
+      end
+
+      if not last_seg and (segments[i+1] ~= "**" or i+1 < #segments) then
+        append({path_sep})
+      end
+    end
+  end
+
+  return patterns
+end
+
+---@private
+--- Implementation of LSP 3.17.0's pattern matching: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#pattern
+--- Modeled after VSCode's implementation: https://github.com/microsoft/vscode/blob/0319eed971719ad48e9093daba9d65a5013ec5ab/src/vs/base/common/glob.ts#L509
```

Comment:
For context, this is about 95% translation from TS to Lua.

I also tested this against VSCode's unit tests and it passes all of them. Is there a way to test this function directly without making it a part of the public API?

Reply:
>  Is there a way to test this function directly without making it a part of the public API?

You'll have to expose it. Usual convention is to prefix it with `_` to indicate it's private (and keep the `@private` annotation to ensure it never gets included in the docs)

Reply:
Just made that change and added a few tests for this. I'll keep adding more once we're happy with the pattern there.

Reply:
> I also tested this against VSCode's unit tests

I expect they are more numerous than what we can support. Is this something we can replicate : how do you run those tests ?

Reply:
I was converting those to Lua by hand, nothing automated. I can't think of a good way we could automatically pull in and run VSCode's tests.

For reference, these are the tests: https://github.com/microsoft/vscode/blob/0319eed971719ad48e9093daba9d65a5013ec5ab/src/vs/base/test/common/glob.test.ts